### DISCUSSION TITLE
How to use sudo ?

### DISCUSSION THREAD

**Discussion by Denis4l** [nushell/nushell#9584] [General]
Hi,
How can we do something like 
`
"/opt/nu-0.78.0-x86_64-unknown-linux-gnu/nu" | sudo save -a /etc/shells
`
for example. In a word how to use sudo ?

**Comment by amtoine** [comment]
afaik, `sudo` will invoke your default shell, which might be `bash` and thus you cannot use `save` in there :thinking: 

i've seen a discussion like this somewhere recently, maybe the doc would benefit having more information about how to use `sudo` :ok_hand:

**Reply by hardfau1t** [reply]
I think sudo looks for the given command in PATH, so shell builtin commands will not work, one way to use save is `"some text" | sudo nu --stdin -c 'save filename'`

**Reply by Denis4l** [reply]
This works, thank you.
So I understand now how to perform a sudo using nu, and more generally, how to interact nu and shell. Good.
Would it be possible to implement something like sudo into nu ?

**Reply by Denis4l** [reply]
...  and as says @amtoine , something about the use of sudo into the doc would be welcome. (.... do it yourself you will say ? :-) )

**Reply by amtoine** [reply]
@Denis4l 
one quick thing that can be done about having a `sudo` in Nushell would be to write a custom command for this, e.g.
```nushell
def sudo [command: string] {
    ^sudo nu --stdin --commands $command
}
```
this is far from perfect but a solution for free :wink: 

about the doc... i think we'd love have a PR to add a little section about using `sudo` in Nushell if it does not already exist :innocent: 
(and if you could do it, as you're interested in that question :smirk: :wink:)

**Reply by Denis4l** [reply]
Interesting, we can define our own command as this ! ... (I haven't check the whole doc)

doc... Heuu, I am curious about all this programming and so on cause I had to code in C++ or java long time ago, but have no skills to do anything. No idea of how to contribute to a doc for example otherwise I would have translate in French and Occitan many things :-) ).

**Reply by Denis4l** [reply]
For the ones whom will look at this thread, defining command : https://www.nushell.sh/book/custom_commands.html

**Reply by i-use-nixos-btw** [reply]
@amtoine i improved on your solution and added typed stdin and stdout via serialization. It can handle all types except closures cause they are not (yet) serialazable in nushell.

```nushell
def sudo [command: closure]: any -> any {
let inp = $in | to nuon
let str_closure = $command | to nuon --serialize | from nuon
let wrapped = $"from nuon | do ($str_closure) | to nuon"
let res = $inp | ^sudo nu --stdin --commands $wrapped
$res | from nuon | return $in }
```

Example: `sudo {ps} | explore`
It would be great to add this to nushell.

**Reply by fdncred** [reply]
closures are serialized with `to nuon --serialize` or `to json --serialize`

**Reply by zaynram** [reply]
@amtoine @i-use-nixos-btw this is a combination of both of your versions which lets you run command strings or closures with --closure. since windows is windows it seems getting full path from (which nu).0.path is more consistent for my setup with wsl.
```nu
# Run a command or closure in an elevated shell
def sudo [command?: string, --closure(-c): closure]: any -> any {
    $in | to nuon | ^sudo (which nu).0.path --stdin --commands (
        if $command != null { $'from nuon | do { ($command) }' }
        else if $closure != null { $'from nuon | do ($closure | to nuon --serialize | from nuon)' }
        else { error make "a command or closure must be provided" }
    )
}
```

**Comment by amtoine** [comment]
> Interesting, we can define our own command as this ! ... (I haven't check the whole doc)
> 
yup we can :smirk: 
this `sudo` custom command is not the best, to be clear, but hopefully it can simplify the times when you need to use `sudo` :crossed_fingers: 

> doc... Heuu, I am curious about all this programming and so on cause I had to code in C++ or java long time ago, but have no skills to do anything. No idea of how to contribute to a doc for example otherwise I would have translate in French and Occitan many things :-) ).
>
if you can do a bit of Git and Github, like the very basics, and use an editor, then you can contribute without a problem :relieved: 
if you need some help, that's no problem, we're here with the @nushell/core-team to guide new contributors through all this and hopefully give all of them a great experience :innocent:
